File: java/src/main/java/parser/util/ParserUtils.java
package parser.util;

import org.apache.commons.lang3.NotImplementedException;
import org.apache.commons.lang3.StringUtils;
import parser.ast.ASTNode;
import parser.ast.Factor;
import parser.ast.FunctionDeclareStmt;

import java.util.ArrayList;
import java.util.LinkedList;

public class ParserUtils {
    // Prefix
    // Postfix
    public static String toPostfixExpression(ASTNode node){

        if(node instanceof Factor) {
            return node.getLexeme().getValue();
        }

        var prts = new ArrayList<String> ();
        for(var child : node.getChildren()) {
            prts.add(toPostfixExpression(child));
        }
        var lexemeStr = node.getLexeme() != null ? node.getLexeme().getValue() : "";
        if(lexemeStr.length() > 0) {
            return StringUtils.join(prts, " ") + " " + lexemeStr;
        } else {
            return StringUtils.join(prts, " ");
        }
    }

    public static String toBFSString(ASTNode root, int max) {

        var queue = new LinkedList<ASTNode>();
        var list = new ArrayList<String>();
        queue.add(root);

        int c = 0;
        while(queue.size() > 0 && c++ < max) {
            var node = queue.poll();
            list.add(node.getLabel());
            for(var child : node.getChildren()) {
                queue.add(child);
            }
        }
        return StringUtils.join(list, " ");
    }
}

File: TinyScript/js/src/parser/util/ParserUtils.js
const Factor = require('../ast/Factor')
const ASTNodeTypes = require('../ast/ASTNodeTypes')
const LinkedList = require('linkedlist')

class ParserUtils {

    
    // postfix :后缀 
    // prefix : 前缀
    // 1 2 3 + + 
    static toPostfixExpression(node) {

        if(node instanceof Factor) {
            return node.getLexeme().getValue()
        }

        const prts = []
        for(const child of node.getChildren()) {
            prts.push(ParserUtils.toPostfixExpression(child));
        }
        const lexemeStr = node.getLexeme() != null ? node.getLexeme().getValue() : "";
        if(lexemeStr.length > 0) {
            return prts.join(" ") + " " + lexemeStr
        } else {
            return prts.join(" ")
        }
    }

    static toBFSString(root, max) {

        const queue = new LinkedList()
        const list = []
        queue.push(root)

        let c = 0
        //     a
        //   b     c
        // d  e  f   g
        while(queue.length > 0 && c++ < max) {
            const node = queue.shift()
            list.push(node.getLabel())
            for(const child of node.getChildren()) {
                queue.push(child);
            }
        }
        return list.join(" ")
    }
}

module.exports = ParserUtils